#292 add RegionTransform.valid_node_types and check in validate

arporter opened the pull request:
An improvement to the functionality of the RegionTrans base class.

arporter commented:
This is now ready for review. It is a small change but since it affects a base class and therefore potentially a lot of functionality I thought I'd break it into a separate PR. @TeranIvy it might be a good one for you to review since it will affect (hopefully help with) your new Extract transformation.

TeranIvy commented:
Happy to take it, @arporter! I think it may not necessarily affect the first round of code review for my PR, but it certainly will the second.

TeranIvy commented:
P.S. It would also be good to do the tidying-up of `RegionTrans` class mentioned in #208, as common validity checks for the list of nodes are repeating in each child class (https://github.com/stfc/PSyclone/issues/208#issuecomment-435934050)

arporter commented on src/psyclone/transformations.py:
```diff
@@ -64,13 +64,21 @@ def __str__(self):
 
 
-# =============================================================================
+@six.add_metaclass(abc.ABCMeta)
```
Yes, that's OK. `ParallelRegionTrans` is itself still abstract.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -64,13 +64,21 @@ def __str__(self):
-    the nodes in the list are in the same order as in the original AST,
-    no node is duplicated, and that all nodes have the same parent.
     '''
+    This abstract class is a base class for all transforms that act on
```
I think either is fine but I've changed it to be on the safe side :-)

arporter commented on src/psyclone/transformations.py:
```diff
@@ -64,13 +64,21 @@ def __str__(self):
-    no node is duplicated, and that all nodes have the same parent.
     '''
+    This abstract class is a base class for all transforms that act on
+    a list of nodes. It gives access to a _validate function that
```
> Also, please add "" at the end of comment lines.

I'm not sure if this is a git-hub markup issue but I don't know what this means?

TeranIvy replied:
I meant "\" but I now see it is not required here.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -79,12 +87,15 @@ class RegionTrans(Transformation):
                 a node is duplicated or the nodes have different parents.
-        '''
+        :raises TransformationError: if any of the type of any of the nodes \
+                                     to be enclosed in the region are not \
```
I've re-worded because it wasn't very good :-)

TeranIvy replied:
Ok. "un-supported" is not consistent with "unsupported" in one of the lines below.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -101,6 +112,16 @@ def _validate(self, node_list):
 
+        # Check that the proposed region contains only supported node types
+        for child in node_list:
+            flat_list = [child] + child.walk(child.children, object)
```
I called it that because I'm converting from a tree structure into a list, a process that is sometimes called 'flattening' (I believe) and also, `list` is a Python built-in.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -1516,6 +1537,11 @@ class OMPParallelTrans(ParallelRegionTrans):
     '''
+    from psyclone import psyGen
+    # The types of node that this transformation can enclose
+    valid_node_types = (psyGen.Loop, psyGen.Kern, psyGen.BuiltIn,
```
I'm not sure I understand but I've added a new test that OMPParallelTrans rejects a node that is not one of these types.

TeranIvy replied:
Thanks!

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -1480,11 +1480,10 @@ def test_acc_data_not_a_schedule():
-        _ = psy.gen
-    assert ("An ACC parallel region must be preceeded by an ACC enter-data "
-            "directive but in invoke_0 this is not the case." in str(err))
+    with pytest.raises(TransformationError) as err:
```
Oh dear, that was sloppy of me. I've broken them into two separate tests now.

TeranIvy replied:
Thanks!

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2236,6 +2262,10 @@ class ProfileRegionTrans(RegionTrans):
     >>> newschedule.view()
 
     '''
+    from psyclone import psyGen, profiler
```
Again, I'm not sure I understand your request? The `_validate()` method is called at line 2316  like so:

    super(ProfileRegionTrans, self)._validate(node_list)

TeranIvy replied:
I missed that, sorry... The coverage is actually fine here so no changes are required.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -101,6 +112,16 @@ def _validate(self, node_list):
+                        "Nodes of type '{0}' cannot be enclosed by a {1} "
+                        "transformation".format(type(item), self.name))
+
+
```
Oh yes, apologies for missing that.

arporter commented:
I've addressed most of the comments although there were a couple that I didn't quite understand. It sounds as though you wanted a test that the `valid_node_types` tuple contains the expected members but since it's just an immutable object I don't think a test for that is worthwhile - e.g. we don't test that lists such as `dynamo0p3.DISCONTINUOUS_FUNCTION_SPACES` have not changed.
Back to @TeranIvy for 2nd review.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -79,12 +87,14 @@ class RegionTrans(Transformation):
                 a node is duplicated or the nodes have different parents.
-        '''
+        :raises TransformationError: if any of the nodes to be enclosed in \
+                the region are of an un-supported type.
```
s/un-supported/unsupported/g. I have seen "unsupported", not so sure about the hyphenated version. Or evene "... are not of a supported type"?

arporter replied:
Oops again - fixed.

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -940,6 +942,23 @@ def test_omp_parallel_do_around_parallel_region():
+def test_omp_region_invalid_node():
+    ''' Check that the OMPParallelTrans transformation rejects nodes
+    of the wrong type. We use an OpenACC directive to trigger this error. '''
+    psy, invoke = get_invoke("single_invoke_three_kernels.f90", API, idx=0)
```
Oops, fixed.

arporter commented:
Both comments addressed. Back to @TeranIvy for 3rd review.
